### ISSUE TITLE
Security configuration problem using 'management.security.enabled=false'

### ISSUE THREAD

**Issue by trisberg** [spring-cloud-stream-app-starters/http#8]
If I create a stream like this:

```
app register --name http --type source --uri maven://org.springframework.cloud.stream.app:http-source-rabbit:1.2.0.RELEASE
app register --name log --type sink --uri maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.2.0.RELEASE
stream create --name test --definition "http --port=8181 | log"
stream deploy --name test --properties "app.http.management.security.enabled=false"
```

Then when posting to the app I get:

```
dataflow:>http post --target http://127.0.0.1:8181/ --data "Hello"
> POST (text/plain) http://127.0.0.1:8181/ Hello
> 403 FORBIDDEN
{
  "path" : "/",
  "error" : "Forbidden",
  "message" : "Could not verify the provided CSRF token because your session was not found.",
  "timestamp" : 1495113416712,
  "status" : 403
}
Error sending data 'Hello' to 'http://127.0.0.1:8181/'
```

**Comment by artembilan** [comment]
Can we have some logs on the matter from the HTTP app, please?

**Comment by trisberg** [comment]
[stdout_0.log.txt](https://github.com/spring-cloud-stream-app-starters/http/files/1010994/stdout_0.log.txt)

**Comment by ilayaperumalg** [comment]
I believe this is because the `http` app has got the basic authentication enabled. What happens when you also pass `security.basic.enabled=false` ? or else, you can find the password for the `user` in the log and use `--username` and `--password` in the http post.

**Comment by artembilan** [comment]
??? That looks like fully unrelated to the HTTP source.
We have here `http.secured` which is `false` by default.

Unclear what may enable security then... 
Something from the SCDF level?

**Comment by trisberg** [comment]
Setting `security.basic.enabled=false` produces the same error. This is a change in behavior from the 1.2.0.RC1 version, so whatever changed has this side effect. The issue came up when I was running the automated acceptance tests, so looking up a password isn't going to work. There is definitely something wrong in the apps configuration. It's very easy to reproduce ;)

**Comment by artembilan** [comment]
@trisberg ,

I have reproduced simply via test-case in the app!
```
@TestPropertySource(properties = {"http.pathPattern=/foo", "management.security.enabled = false"})
```
and the result is like:
```
{timestamp=1495120201023, status=403, error=Forbidden, message=Could not verify the provided CSRF token because your session was not found., path=/foo}
```

So, that looks like `"management.security.enabled = false` breaks Spring Boot auto-config conditions somehow...

IMO having:
```
@EnableWebSecurity
protected static class DisableSecurityConfiguration { }
```
in your code should not be affected by those `management.security.enabled` and `security.basic.enabled` options. Since I have my own definition it is already my responsibility to configure everything I need.

Investigating...

Should we loop it with Boot guys, though?

**Comment by artembilan** [comment]
Just in case you would like to know more weird info.
When I do `management.security.enabled = true`, my test works. 😕

**Comment by artembilan** [comment]
OK. Here is what I have. 
(Correct me if I am missing something).

Spring Boot has this:
```
@Configuration
@ConditionalOnMissingBean({ ManagementWebSecurityConfigurerAdapter.class })
@ConditionalOnProperty(prefix = "management.security", name = "enabled", matchIfMissing = true)
@Order(ManagementServerProperties.BASIC_AUTH_ORDER)
protected static class ManagementWebSecurityConfigurerAdapter
			extends WebSecurityConfigurerAdapter {
```
So, if we don't have `management.security.enabled` or it is `true`, this configurer is switched on.
The `WebSecurityConfiguration` (a result of the `@EnableWebSecurity`) does then this:
```
@Bean(name = AbstractSecurityWebApplicationInitializer.DEFAULT_FILTER_NAME)
public Filter springSecurityFilterChain() throws Exception {
	boolean hasConfigurers = webSecurityConfigurers != null
			&& !webSecurityConfigurers.isEmpty();
	if (!hasConfigurers) {
		WebSecurityConfigurerAdapter adapter = objectObjectPostProcessor
				.postProcess(new WebSecurityConfigurerAdapter() {
				});
		webSecurity.apply(adapter);
	}
	return webSecurity.build();
}
```
If we have those custom `webSecurityConfigurers`, the default `WebSecurityConfigurerAdapter` isn't applied.
But if we do `management.security.enabled = false`, that `ManagementWebSecurityConfigurerAdapter` isn't matched and we don't have custom `webSecurityConfigurers` at all. Therefore default Security schema is applied like:
```
if (!disableDefaults) {
			// @formatter:off
			http
				.csrf().and()
				.addFilter(new WebAsyncManagerIntegrationFilter())
				.exceptionHandling().and()
				.headers().and()
				.sessionManagement().and()
				.securityContext().and()
				.requestCache().and()
				.anonymous().and()
				.servletApi().and()
				.apply(new DefaultLoginPageConfigurer<HttpSecurity>()).and()
				.logout();
			// @formatter:on
			ClassLoader classLoader = this.context.getClassLoader();
			List<AbstractHttpConfigurer> defaultHttpConfigurers =
					SpringFactoriesLoader.loadFactories(AbstractHttpConfigurer.class, classLoader);

			for(AbstractHttpConfigurer configurer : defaultHttpConfigurers) {
				http.apply(configurer);
			}
}
```
That's how we have that `CsrfFilter` activated and end up with the:
```
public MissingCsrfTokenException(String actualToken) {
	super("Could not verify the provided CSRF token because your session was not found.");
}
```
because we don't provide that CSRF token.

Does it make sense to you?

Should I think how to overcome?
For example provide our custom `WebSecurityConfigurerAdapter` inside that `DisableSecurityConfiguration` to really disable.

When `http.secured = true` that one will be ignored and everything what you want from Spring Boot security will be possible.

**Comment by artembilan** [comment]
Another observation. Since we configure `@EnableWebSecurity`, the `SpringBootWebSecurityConfiguration` doesn't match, therefore `security.basic.enabled=false` does not have effect because we can't step into:
```
@Configuration
@ConditionalOnProperty(prefix = "security.basic", name = "enabled", havingValue = "false")
@Order(SecurityProperties.BASIC_AUTH_ORDER)
protected static class ApplicationNoWebSecurityConfigurerAdapter
	extends WebSecurityConfigurerAdapter {
```

So. from here it sounds like the custom `@EnableWebSecurity` config like it is recommended by Boot itself isn't robust solution: http://docs.spring.io/spring-boot/docs/1.5.3.RELEASE/reference/html/howto-security.html#howto-switch-off-spring-boot-security-configuration.

Maybe we should bug Boot on the matter? Like `ManagementWebSecurityAutoConfiguration` should be marked with the `@ConditionalOnMissingBean(WebSecurityConfiguration.class)` as well 😄

**Comment by artembilan** [comment]
OK. I went ahead and fixed it the way I see: #9 
That might be just as a workaround until proper solution in Boot. (or we will figure out anything else).
But at the same time we have to unblock the feature somehow...
